pep-0622.rst

```diff
@@ -21,103 +21,254 @@ Resolution:
-it was discovered that ``isinstance()`` is the second most called builtin
-function (after ``len()``). Even taking into account builtin classes, it is
-still in the top ten. Most of such calls are followed by specific attribute
-access.
-
-There are two possible conclusions that can be drawn from this information:
-
-* Handling of heterogeneous data (i.e. situations where a variable can take
-  values of multiple types) is common in real world code.
-
-* Python doesn't have expressive ways of destructuring object data (i.e.
-  separating the content of an object into multiple variables).
-
-This is in contrast with the opposite sides of both aspects:
-
-* Its success in the numeric world indicates that Python is good when
-  working with homogeneous data. It also has builtin support for homogeneous
-  data structures such as e.g. lists and arrays, and semantic constructs such
-  as iterators and generators.
-
-* Python is expressive and flexible at constructing objects. It has syntactic
-  support for collection literals and comprehensions. Custom objects can be
-  created using positional and keyword calls that are customized by special
-  ``__init__()`` method.
-
-This PEP aims at improving the support for destructuring heterogeneous data
-by adding a dedicated syntactic support for it in the form of pattern matching.
-On a very high level it is similar to regular expressions, but instead of
-matching strings, it will be possible to match arbitrary Python objects.
-
-We believe this will improve both readability and reliability of relevant code.
-To illustrate the readability improvement, let us consider an actual example
-from the Python standard library::
-
-  def is_tuple(node):
-      if isinstance(node, Node) and node.children == [LParen(), RParen()]:
-          return True
-      return (isinstance(node, Node)
-              and len(node.children) == 3
-              and isinstance(node.children[0], Leaf)
-              and isinstance(node.children[1], Node)
-              and isinstance(node.children[2], Leaf)
-              and node.children[0].value == "("
-              and node.children[2].value == ")")
-
-With the syntax proposed in this PEP it can be rewritten as below. Note that
-the proposed code will work without any modifications to the definition of
-``Node`` and other classes here::
-
-  def is_tuple(node: Node) -> bool:
-      match node:
-          case Node(children=[LParen(), RParen()]):
-              return True
-          case Node(children=[Leaf(value="("), Node(), Leaf(value=")")]):
-              return True
-          case _:
-              return False
+Python programs frequently need to handle data which varies in type,
+presence of attributes/keys, or number of elements. Typical examples
+are operating on nodes of a mixed structure like an AST, handling UI
+events of different types, processing structured input (like
+structured files or network messages), or “parsing” arguments for a
```

Comment:
"Structure" repeated:

```suggestion
files or network messages), or “parsing” arguments for a
```

Reply:
Intentional, given that many files are not structured input but just a stream of bytes